sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/data/BeamFnDataGrpcMultiplexer.java

```diff
@@ -123,43 +141,142 @@ public void close() {
+        if (!consumer.isDone()) {
+          LOG.debug(
+              "Received data for key {} without consumer ready. "
+                  + "Waiting for consumer to be registered.",
+              key);
+        }
+        consumer.get().accept(data);
+        if (isFinalizationData(data)) {
+          consumers.remove(key);
+        }
+        /*
+         * TODO: On failure we should fail any bundles that were impacted eagerly
+         * instead of relying on the Runner harness to do all the failure handling.
+         */
+      } catch (ExecutionException | InterruptedException e) {
+        LOG.error(
+            "Client interrupted during handling of data for instruction {} and target {}",
+            data.getInstructionReference(),
+            data.getTarget(),
+            e);
+        outboundObserver.onError(e);
+      } catch (RuntimeException e) {
+        LOG.error(
+            "Client failed to handle data for instruction {} and target {}",
+            data.getInstructionReference(),
+            data.getTarget(),
+            e);
+        outboundObserver.onError(e);
+      }
+    }
+
+    private boolean IsQueueingEnabled() {
+      return instructionIdToQueue != null;
+    }
+
+
+    // TODO please help reveiw the concurrency aspect here. Questions:
+    // - Safe use of ConcurrentHashMap?
+    // - Unnecessary use of ConcurrentHashMap?
+    // - Is there some code that can run concurrently now, which was not before?
     @Override
     public void onNext(BeamFnApi.Elements value) {
       for (BeamFnApi.Elements.Data data : value.getDataList()) {
         try {
-          LogicalEndpoint key =
-              LogicalEndpoint.of(data.getInstructionReference(), data.getTarget());
-          CompletableFuture<Consumer<BeamFnApi.Elements.Data>> consumer = receiverFuture(key);
-          if (!consumer.isDone()) {
-            LOG.debug(
-                "Received data for key {} without consumer ready. "
-                    + "Waiting for consumer to be registered.",
-                key);
-          }
-          consumer.get().accept(data);
-          if (data.getData().isEmpty()) {
-            consumers.remove(key);
+          if (IsQueueingEnabled()) {
+            // TODO Is there a cleaner way do to this?
+            // I was trying to avoid always doing multiple lookups, and always creating
+            // a new instance on the heap
+            SynchronousQueue<Data> queue = instructionIdToQueue.get(data.getInstructionReference());
+            if (queue == null) {
+              instructionIdToQueue.putIfAbsent(
+                  data.getInstructionReference(), new SynchronousQueue<Data>());
+              // Lookup the queue again incase there was a race and another thread added one.
+              queue = instructionIdToQueue.get(data.getInstructionReference());
+            }
+
+            // TODO, how do we make this throw an interrupted exception? And when should we
```

Comment:
You could build a poll loop using [Queue#offer(element, time)](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/SynchronousQueue.html#offer(E,%20long,%20java.util.concurrent.TimeUnit))

Each time you would need to check to see if the "consumer" is dead. This would require you to swap to have a consumer that exposes whether its done or not.

Reply:
I think what I did makes sense